02-intermediate-ruby/testing-with-minitest.md

```diff
@@ -0,0 +1,419 @@
+
+end
+```
+
+Now `ruby bill_spec.rb` gets us:
+
+```bash
+Run options: --seed 32657
+
+# Running:
+
+..
+
+Finished in 0.001089s, 918.2736 runs/s, 918.2736 assertions/s.
+
+1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
+
+Finished in 0.454822s, 2.1987 runs/s, 2.1987 assertions/s.
+1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
+
+```
+
+Now we have our first **green/passing** test.
+
+There are a [number of expectations](http://mattsears.com/articles/2011/12/10/minitest-quick-reference/) in Minitest beyond the `must_equal` method.
+
+**Exercise** Look at the list of expectations above and revise the `bill.class.must_equal Bill` to use the `must_be_instance_of` expectation instead.
+
+
+#### A Word on Parentheses
+
+In the code above, we are calling `must_equal` without using parentheses `bill.class.must_equal Bill`.  Ruby doesn't FORCE you to put parentheses around a method's arguments but [the community-driven style guidelines](https://github.com/bbatsov/ruby-style-guide#method-invocation-parens) suggest that it's good coding style to put parentheses around method arguments **except** for methods which are part of an internal Domain Specific Language (DSL), or basically the syntax of some kind of framework like... Minitest.
+
+So you shouldn't put parentheses around the arguments to `must` method arguments, but you **should** around your own methods.
+
+### Adding Functionality
+
+We are going to require that Bill must be instantiated with a list (array) of hashes.  Each item will have a name and price.  So we will revise the test to look like this:
+
+```ruby
+# bill_spec.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'bill'
+
+describe "bill" do
+  it "can be created" do
+    bill = Bill.new([4.75, 8.75])
+    # the class of @bill should be Bill
+    bill.class.must_equal Bill
+  end
+end
+```
+
+#### Our First Edge Cases
+
+You may remember from the [Introduction to Automated Testing](../00-programming-fundamentals/intro-to-automated-tests.md) that we also have an edge-case when the list is empty.  As designers we decide what to do when:
+
+1. A `Bill` is constructed without an argument
+1. A `Bill` is constructed with a non-array argument
+1. Or Bill.new is called an empty list
+
+We will choose to **raise an error**.  By doing this we can insure that no `Bill` is created in an invalid state.
+
+So we will add another `it` block and use the `must_raise` expectation.
+
+```ruby
+it "will raise an error if created without a list of prices" do
+  proc {
+    bill = Bill.new
+  }.must_raise ArgumentError
+end
+```
+
+**Question**: The code above tests for the first situation.  How can you write a tests for the others?  You can find a solution [here](https://github.com/AdaGold/bill_calculator/commit/96f00a74e60d4e4246d0c8a10dfe66dea89cfd6d)
+
+#### How to make it pass?
+
+Now we need to write code to make the tests pass.  Raising an error is new for us right now, so lets look at how we can do so:
+
```

Comment:
I have a few comments about this section:

1. I am nervous about introducing `raise` for the first time here - this is already going to be an overwhelming amount of new stuff for the students, and I think this dilutes our focus
1. You code block above, where you call `Bill.new` with no arguments, is testing built-in ruby functionality. There's not a line of code we could remove or change to make it not raise an argument error, so I would argue it's not something we should be testing.
1. Checking the class of an argument is a classic anti-pattern in dynamically typed languages, and is not a behavior we should be modeling for students. See [this article](https://mikewilliamson.wordpress.com/2013/01/16/duck-typing-in-ruby-thoughts-on-checking-argument-types/) or Metz ch 5.

One way to resolve these issues might be to:

- Re-write the test above to pass in an empty array, and don't add tests for invalid or too few arguments
- Change the `Bill` implementation to be OK with an empty array

The test will pass without us having to add any code, which is not ideal, but I still think it will make this section flow a little smoother.

Reply:
In C8, raising an error didn't seem to cause much confusion, and it later set the stage for how to use `rescue`.  There are many confusing things about testing, but raise didn't seem to be one of them.  

I'll take it out, but we'll need another introduction later, maybe just with the livecode covering cards to introduce it.